discovery/src/main/java/org/wildfly/extension/discovery/AggregateProviderAddHandler.java

```diff
@@ -0,0 +1,90 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.wildfly.extension.discovery;
+
+import java.util.List;
+
+import org.jboss.as.controller.AbstractAddStepHandler;
+import org.jboss.as.controller.AttributeDefinition;
+import org.jboss.as.controller.OperationContext;
+import org.jboss.as.controller.OperationFailedException;
+import org.jboss.as.controller.capability.RuntimeCapability;
+import org.jboss.as.controller.operations.global.WriteAttributeHandler;
+import org.jboss.as.controller.registry.Resource;
+import org.jboss.dmr.ModelNode;
+import org.wildfly.discovery.impl.AggregateDiscoveryProvider;
+import org.wildfly.discovery.impl.MutableDiscoveryProvider;
+import org.wildfly.discovery.spi.DiscoveryProvider;
+
+/**
+ * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
+ */
+class AggregateProviderAddHandler extends AbstractAddStepHandler {
+    private static final AggregateProviderAddHandler INSTANCE = new AggregateProviderAddHandler();
+
+    static AggregateProviderAddHandler getInstance() {
+        return INSTANCE;
+    }
+
+    private AggregateProviderAddHandler() {
+        super(new Parameters().addAttribute(AggregateProviderDefinition.PROVIDER_NAMES));
+    }
+
+    public void execute(final OperationContext context, final ModelNode operation) throws OperationFailedException {
+        context.registerCapability(
+            RuntimeCapability.Builder
+                .of(DiscoveryExtension.DISCOVERY_PROVIDER_CAPABILITY, true, new MutableDiscoveryProvider())
+                .build().fromBaseCapability(context.getCurrentAddressValue()));
+        super.execute(context, operation);
+    }
+
+    protected void recordCapabilitiesAndRequirements(final OperationContext context, final ModelNode operation, final Resource resource) throws OperationFailedException {
+        final AttributeDefinition ad = AggregateProviderDefinition.PROVIDER_NAMES;
+        ad.addCapabilityRequirements(context, resource.getModel().get(ad.getName()));
+    }
+
+    static void modifyRegistrationModel(OperationContext context, ModelNode op) throws OperationFailedException {
+        WriteAttributeHandler.INSTANCE.execute(context, op);
+        context.addStep(op, AggregateProviderAddHandler::modifyRegistration, OperationContext.Stage.RUNTIME);
+    }
+
+    static void modifyRegistration(OperationContext context, ModelNode op) throws OperationFailedException {
+        final MutableDiscoveryProvider mutableDiscoveryProvider = context.getCapabilityRuntimeAPI(DiscoveryExtension.DISCOVERY_PROVIDER_CAPABILITY, context.getCurrentAddressValue(), MutableDiscoveryProvider.class);
+        if (op.hasDefined(DiscoveryExtension.PROVIDERS)) {
```

Comment:
This executes in Stage.RUNTIME so generally you don't want to use param values from 'op' here. By now the param values will have been stored in the model. So get the model and use that where you're using 'op' here:

ModelNode model = context.readResource(PathAddress.EMPTY_ADDRESS).getModel();

This ensures that if people do weird composite operations, wit multiple steps modifying the model, when your RUNTIME steps runs it is working with the end result of that.

Ah, actually you have to use the model here regardless, since in a write-attribute, the 'op' would be

{
 "operation=>"write-attribute",
 "address"=>[.....],
 "name"=>"providers",
 "value"=>["a","b","c"]
}

Doing op.get("providers") on that is not going to give you ["a","b","c"].